### ISSUE TITLE
Add support for Android 5.0

### ISSUE THREAD

**Issue by dtr2** [xroche/coffeecatch#15]
There's no "libcorkscrew" in Android 5.0. Instead, there's a C++-based "libbacktrace.so".
here's a sample code to use it:
# include "backtrace/Backtrace.h"

```
int btdump(int sig, struct siginfo * info, void * ctx) {

    //Backtrace* t = Backtrace::Create(getpid(),0,NULL );
    void* libbacktrace = dlopen( "libbacktrace.so", RTLD_LOCAL );
    if ( !libbacktrace ) 
        return -1;
    Backtrace* (*create)(int,int,void*);
    *(void**)&create = dlsym( libbacktrace, "_ZN9Backtrace6CreateEiiP12BacktraceMap" );
    if ( !create ) 
        return -1;
    Backtrace* t = create(getpid(), 0, NULL );
    if ( !t ) 
        return -1;

    int ret = t->Unwind(0, (ucontext*)ctx);
    if ( !ret ) {
        return -1;
    }
    int count = t->NumFrames();

    DUMP( "backtrace:" );
    for ( int i=0; i<count; i++ ) {
        std::string line = t->FormatFrameData(i);
        DUMP( "\t%s",line.c_str());

    }
    return 0;
}
```

**Comment by zammbi** [comment]
Any update on this?

**Comment by Jayatubi** [comment]
How about the "libunwind.so" which shipped with 5.0?

**Comment by dtr2** [comment]
libunwind.so is there.

-- dror.

**Comment by xroche** [comment]
Yep, I'm lagging a bit, but anyway unwind support is already amost there in coffeecatch, I probably just need to dynamically load it.

**Comment by Jayatubi** [comment]
Is there any way to make libunwind to work with sigaction? All the samples I could found on the internet are about the current stack only.

**Comment by Jayatubi** [comment]
And I've made the libcorkscrew works on Android 5 by extract the code from AOSP and compile it statically into my project. It works well by I saw some issues about libcorkscrew here: https://groups.google.com/d/msg/android-platform/VKlX7sNEV10/2XQNqy5awjkJ

**Comment by xroche** [comment]
Darn, **of course** the Android libunwind.so is not the "classical" gcc-style libunwind (*_Unwind_Backtrace()* et al.) but with a totally different API. Lollipop is a definitely real pain in the ***. 
> *"Backward compatibility ? Who cares!"*

Stay tuned, I'll have to dig in this new exotic API!

[rant]

I'm still puzzled that the inane libbacktrace API has been pushed as the new official one
  * C++: unusable for C-only code, require uncool non-standard symbol mangling for dynamic load
  * Use of C++ objects absolutely everywhere (hey! this API will be used in case of crashes, and possibly within signal handlers, so let's allocate dozens of C++ objects because it is always safe to do that in a signal-safe way! yes, trust me!)
  * Use of std::string (you wanted to use this API in pure C ? Nope!) even to store symbol names

[/rant]

**Comment by dtr2** [comment]
Also, if you look at stack traces on L devices, you'll see that it fails to
unwind the stack in many cases. usually 2 lines of stack (or one) and
that's all
Also, by default they have the processes un-dumpable. In order to collect
stack traces of my code, I add to the program:
prctl(PR_SET_DUMPABLE, 1);
without which, the debuggerd can't do anything.

-- dror.

**Comment by xroche** [comment]
> Also, if you look at stack traces on L devices, you'll see that it fails to unwind the stack in many cases. usually 2 lines of stack (or one) and that's all

Even with code built with `-funwind-tables` ?

**Comment by dtr2** [comment]
Didn't check. I'm talking about stack traces of arbitrary code, not code
that was compiled specifically to allow backtraces...


-- dror.

**Comment by VioletGiraffe** [comment]
Not sure if it's anything new to you, but this code works on my Android 5.0 Nexus: http://stackoverflow.com/a/28858941/634821

Closer to the topic: I can't make use of the code because coffecatch fails to catch the signal in the first place - `COFFEE_CATCH` never gets called. Any progress on this?

**Comment by anyshu** [comment]
Hi dtr2, 
I used your sample code to catch call stack during crash. And an error occurred which "W/libbacktrace( 7610): virtual bool BacktraceThread::Unwind(size_t, ucontext_t*): tgkill 0 failed: Invalid argument", could you help to check what's problem? 
I tested on Nexus 5 Android 5.02 and 5.1.

Thanks.

**Comment by ennetys** [comment]
Hi dtr2,
I work on Android 5.1.1. I have one question?
how to use the function "int btdump(int sig, struct siginfo * info, void * ctx)"?
Contretely, how can I get the parameter "ctx"? In the ucontext.h file(android platform), I do not find the function getcontext(..). if ctx == null, failure will be returned when calling UnWind function.

This puzzle me a lot.
Thanks

**Comment by dtr2** [comment]
Sorry didn't see your email earlier.

I need to see if i can test it. I no longer work on this jni based
project...

**Comment by xroche** [comment]
Android 5+ is handled through the libunwind backtrace path. Closing as fixed.
